### ISSUE TITLE
Themes: Make it easy to create beautiful plots with hvPlot

### ISSUE THREAD

**Issue by MarcSkovMadsen** [holoviz/hvplot#834]
Currently there is no easy (i.e. one-line of code) to make your hvPlot beautiful. 

And (in my opinion) the existing theme has not been attractive for a decade. The colors, fonts and other details needs a modernization. I believe the lots of users will drop a plotting library if they cannot easily create great looking plots.

![hvplot](https://user-images.githubusercontent.com/42288570/184523596-9c727b54-8bc8-4771-b6e3-dd5cd147d5ad.png)

I would like a oneliner that can see the theme used for hvPlot.

```python
.set_theme("made_up")
```

In my dream world the same theme `made_up` theme would work across backends and across hvplot, HoloViews and Panel.

I believe especially having a beautiful theme for Matplotlib+Bokeh backends in `light` + `dark` mode would be really, really valuable. As there are already many Matplotlib themes out there it might be possible to convert one of these to a Bokeh theme. The existing bokeh themes are ugly - don't use them.

I would also like the ability to specify one accent color similarly to the Fast templates, potentially a color palette and font to enable users to easily use their organizations style. For private fonts I would need to the ability to refererence a css style sheet in hvPlot.

Personally I would like this for work and my private life - for example whenever I demo hvPlot and HoloViews and try to communicate how powerful it is. Its just difficult because you cannot get over the initial hurdle of the output not looking awesome by default. People just don't see the rest.

A simple minimal solution might to be Gallery notebook examples of how to style each backend.

While we are at it the Bokeh default tool should also be changed to zoom. I see users being confused all the time because it is not the default. And double-click should zoom out. And there is no need to display hover tool icons. This would all significantly improve the user experience.

**Comment by MarcSkovMadsen** [comment]
There is some discussion with useful information here. But you can see how painful it is and how you have to use Panel to get an end to end solution supporting custom fonts.

https://discourse.holoviz.org/t/how-do-i-use-custom-font-in-holoviews/303/11

**Comment by MarcSkovMadsen** [comment]
For example I can easily change the color map for bar charts. But the hover color and color user for the selected tools is still blue.

It does not look appealing and as something you would present to any other than you self and your data colleagues.

![image](https://user-images.githubusercontent.com/42288570/185799643-5587eb53-b1e3-420d-ba5e-96d4a5bc2fea.png)

Being able to easily select a better theme including a few variables like font, accent color and color map would make a world of a difference. 

We have that for the Panel FastListTemplate. Must be possible to carve out.

**Comment by maximlt** [comment]
> And (in my opinion) the existing theme has not been attractive for a decade. 

I assume you're talking about Bokeh since hvPlot is only a few years old. Did you open a feature request on Bokeh's Github to improve its theme? That would benefit a lot of people, not just hvPlot's users.

> Its just difficult because you cannot get over the initial hurdle of the output not looking awesome by default. People just don't see the rest.

That is not really the experience I've got. For instance, Matplotib plots aren't particularly beautiful by default, yet there's millions of Matplotlib plots out there.

I think the source of your tension using hvPlot is that you are using it just to make beautiful plots. I think hvPlot and HoloViews should be improved to make plots that are more beautiful and publication-ready, and that the documentation should be improved to show how to do that. But what they excel at is data exploration, which is the step in a data analysis workflow where you're trying to understand the data and should not (really, you should not) focus on how things look, as it can only distract you from your goal.

@jlstevens is there a way in HoloViews to define themes?

**Comment by jlstevens** [comment]
For the most part, HoloViews simply exposes whatever theming support exists in the plotting backend. For instance, Bokeh plots can have bokeh themes applied.

Of course, HoloViews has some of its own defaults that can be overridden as desired - if you wanted to you could collect yours preferences together and override some global style/plotting options. As a whole, we haven't really done this much though, but of course if someone were to make a nice alternative theme, we would be very happy to consider making it available to everyone!

**Comment by MarcSkovMadsen** [comment]
Its actually not the theme.

It requires css as well for which i dont know a mechanism to inject.

And i believe it should be possible to quickly share the resultat of exploration in a nice looking and functioning way with colleges, managers and other stakeholders.

Furthermore we also need the web site to be nice looking to appeal to users. And the same when used in blogs, social media and when answering users on discourse.

Right now the web site reference examples looks like exploration. There is no easy way to get to an end result.

**Comment by MarcSkovMadsen** [comment]
Still needed. I tried to create a reference bar chart example for an AI Agent Skill in https://github.com/MarcSkovMadsen/holoviz-mcp/pull/73 with some inspiration from https://www.atlassian.com/data/charts/bar-chart-complete-guide.

<img width="1732" height="862" alt="Image" src="https://github.com/user-attachments/assets/4d4be4b6-8e71-413c-8f25-08929379aacc" />

It just takes a lot of code and some things don't work like `tools`, are hard to get working (why is both `hover_cols` and `hover_tooltips` needed? Why are `fields` used to rename the columns instead of just how they are presented?) or are on the Bokeh level. 

<details><summary>Code</summary>

```python
# ============================================================================
# Publication-Quality Bar Chart - hvPlot Best Practices Example
# ============================================================================
# Demonstrates:
# - Data extraction, transformation, and visualization separation
# - Custom Bokeh themes for consistent styling
# - Interactive tooltips with formatted data
# - Text annotations on bars
# - Professional fonts, grids, and axis formatting
# - Panel integration for web serving
# ============================================================================

import hvplot.pandas  # noqa: F401
import panel as pn
import hvsampledata
from bokeh.models.formatters import NumeralTickFormatter
from bokeh.themes import Theme
import holoviews as hv


# ============================================================================
# THEME SETUP - Define global styling
# ============================================================================

def create_bokeh_theme(font_family='Roboto'):
    """Create custom theme with specified font. Default: Roboto"""
    return Theme(json={
        'attrs': {
            'Title': {
                'text_font': font_family,
                'text_font_size': '16pt',
                'text_font_style': 'bold'
            },
            'Axis': {
                'axis_label_text_font': font_family,
                'axis_label_text_font_size': '12pt',
                'axis_label_text_font_style': 'bold',
                'major_label_text_font': font_family,
                'major_label_text_font_size': '10pt',
                'major_tick_line_color': None,  # Remove tick marks
                'minor_tick_line_color': None
            },
            'Grid': {
                'grid_line_color': '#e0e0e0',
                'grid_line_alpha': 0.5
            },
            'Plot': {
                'background_fill_color': '#fafafa',
                'border_fill_color': '#fafafa'
            },
            'Legend': {
                'label_text_font': font_family,
                'label_text_font_size': '10pt'
            }
        }
    })

# Apply theme globally - affects all plots
hv.renderer('bokeh').theme = create_bokeh_theme('Helvetica')


# ============================================================================
# DATA PIPELINE - Separate extraction, transformation, and plotting
# ============================================================================

def get_earthquake_data():
    """Extract raw earthquake data from sample dataset"""
    return hvsampledata.earthquakes("pandas")


def aggregate_by_magnitude(earthquake_data):
    """Transform: Group earthquakes by magnitude class with statistics"""
    
    # Aggregate: count events and calculate average depth per magnitude class
    aggregated = (
        earthquake_data
        .groupby('mag_class', observed=True)
        .agg({'mag': 'count', 'depth': 'mean'})
        .reset_index()
        .rename(columns={'mag': 'event_count', 'depth': 'avg_depth'})
        .sort_values('event_count', ascending=False)
    )

    # Add percentage column for tooltips
    aggregated['percentage'] = (
        aggregated['event_count'] / aggregated['event_count'].sum() * 100
    )

    return aggregated


def create_bar_chart(aggregated_data):
    """Create publication-quality bar chart with labels and tooltips"""
    
    # Main bar chart with professional styling
    bar_chart = aggregated_data.hvplot.bar(
        x='mag_class',
        y='event_count',
        
        # Titles and labels
        title='Earthquake Distribution by Magnitude',
        xlabel='Magnitude',
        ylabel='Number of Events',
        
        # Visual styling
        color='#3b7ea1',           # Professional blue
        line_color=None,            # Remove bar borders
        
        # Interactivity
        hover_cols = ["mag_class", "event_count", "percentage", "avg_depth"],
        hover_tooltips=[
            ('Magnitude', '@mag_class'),
            ('Events', '@event_count{0,0}'),      # Format: 1,234
            ('Percentage', '@percentage{0 a}%'), # Format: 45.7%
            ('Avg Depth', '@avg_depth{0f} km')  # Format: 99 km
        ],
        
        tools=['hover', 'save'],
                
        # Axis formatting - use NumeralTickFormatter for readability
        yformatter=NumeralTickFormatter(format='0a'),  # 1k, 1.5M, etc.
        autohide_toolbar=True,
        # Grid and layout
        legend='top_right',
        # Backend customization
        backend_opts={
            'plot.xgrid.visible': False,           # Only horizontal grid lines
            'plot.ygrid.grid_line_color': "black",
            'plot.ygrid.grid_line_alpha': 0.1,
            'plot.title.text_font_style': 'bold',
            'plot.xaxis.axis_label_text_font_style': 'bold',
            'plot.yaxis.axis_label_text_font_style': 'bold',
            'plot.min_border_left': 80,            # Add padding on left (for y-axis label)
            'plot.min_border_bottom': 80,          # Add padding on bottom (for x-axis label)
            'plot.min_border_right': 30,           # Add padding on right
            'plot.min_border_top': 80,             # Add padding on top
        }
    )

    # Add text labels above bars
    labels_data = aggregated_data.copy()
    labels_data['label_y'] = labels_data['event_count'] + 20  # Offset above bars
    
    text_labels = labels_data.hvplot.labels(
        x='mag_class',
        y='label_y',
        text='event_count',
        text_baseline='bottom',
        text_font_size='11pt',
        text_font_style='bold',
        text_color='#333333',
        hover=False
    )

    # Overlay: bar chart * text labels
    return bar_chart * text_labels


def create_plot():
    """Main function: Extract → Transform → Plot"""
    # Extract: Get raw data
    earthquake_data = get_earthquake_data()
    
    # Transform: Aggregate and calculate statistics
    aggregated = aggregate_by_magnitude(earthquake_data)
    
    # Visualize: Create publication-quality chart
    chart = create_bar_chart(aggregated)
    
    return chart


# ============================================================================
# PANEL APP SETUP
# ============================================================================

# Serve the chart when running with Panel
if pn.state.served:
    # Load Panel JavaScript extensions
    pn.extension()
    
    # Apply custom Bokeh theme (override the global theme)
    theme = create_bokeh_theme()
    hv.renderer('bokeh').theme = theme
    
    # Create and serve the chart
    chart = create_plot()
    pn.panel(chart, sizing_mode="stretch_both", margin=25).servable()
```
<details>

**Comment by ahuang11** [comment]
I think using a theme makes this a lot easier:

`hv.renderer("bokeh").theme = "light_minimal"`
https://docs.bokeh.org/en/latest/docs/reference/themes.html

**Comment by MarcSkovMadsen** [comment]
The request is still a one-liner to get beautiful plots. The above was just another example to motivate that this is still needed.

**Comment by ahuang11** [comment]
I can empathize hvPlot / holoviews / bokeh plots are not subjectively "beautiful", which is why I created most of the [themes available in bokeh](https://docs.bokeh.org/en/latest/docs/user_guide/styling/themes.html) ;)

However, as you mentioned:

> bokeh does not provide any explanation of how a Theme works and what you can expect.

And I also agree with that so: https://github.com/bokeh/bokeh/issues/14762

> I don't know what to do with these kinds of comments. It'd be great if you could try to give more constructive feedback. Perhaps, you could compare hvPlot's defaults with another library's defaults (e.g., Plotly), and show how the defaults of the other library are superior and more sensible? 

I think the biggest problem with bokeh defaults is the "chart noise", adding to the cognitive noise and distracting the main message. The two big ones that I immediately notice (and get bothered by) are the minor ticks and italics in the x/y labels.

Here's before / after with `light_minimal` theme

<img width="715" height="455" alt="Image" src="https://github.com/user-attachments/assets/4344173e-af64-44ea-b63c-1c253fa14825" />

<img width="721" height="435" alt="Image" src="https://github.com/user-attachments/assets/36567e0b-6020-4ac2-bea8-571e9c06408c" />

It's not perfect; looking back at it (a few years later), the colors of light_minimal should be darker and the title should be larger, but disregarding that, I *think* it already looks a lot much better and solid than the default.

As I [previously mentioned in the roadmap PR](https://github.com/holoviz/hvplot/pull/1694#pullrequestreview-3541768248), and now hijacking this issue:

> Many users reach for Plotly or Seaborn because their defaults are immediately more visually appealing (and in Seaborn’s case, offer built-in statistical plotting). We could also document how to better integrate with websites, dashboards, and publications.

Adding a few more reference links I've found since then:
- https://www.linkedin.com/feed/update/urn:li:activity:7406682331547062272/
- https://www.react-graph-gallery.com/example/line-chart-design-improvement
- https://dominicroye.github.io/blog/2025-12-14-broken-charts/

(Yan Holtz has been sharing many great resources on improving chart UI/UX, check out his [feed](https://www.linkedin.com/in/yan-holtz-2477534a/))

**Comment by maximlt** [comment]
> And I also agree with that so: [bokeh/bokeh#14762](https://github.com/bokeh/bokeh/issues/14762)

Thanks for opening that issue! It seems to me @MarcSkovMadsen was more after docs (he said "explanations") but I'll let him clarify his ask if he wants.

**Comment by maximlt** [comment]
@ahuang11 do you think there's a chance improvements could be made directly to Bokeh's default theme? @mattpap mentioned there's some theming work planned for the CZI project:

> We may be able to do that as a part of CZI theming work.

To try to avoid this.

<img width="1000" height="567" alt="Image" src="https://github.com/user-attachments/assets/d71d9976-725d-42f3-975e-6a919db59b84" />

Although, @MarcSkovMadsen did mention that a better theme wouldn't fully meet its needs (I didn't really understand why).

For inspiration, I found this pretty cool [page](https://vega.github.io/vega-themes/) from Vega, where you can switch themes and see 8 different example plots. They do have "business" themes like Excel, Google Charts, Power BI.

**Comment by ahuang11** [comment]
Yes if Bokeh docs had something similar to that, that'd be great!

**Comment by jbednar** [comment]
> Personally I find it more distracting to have UI components show and hide whenever mouse cursor goes into and out of plots. 

I agree that hiding and revealing components can be distracting, but on balance I strongly prefer to have the toolbar hidden by default so that a screenshot or printout of the page will not show the (now useless) tools, only the data itself.  Similarly for someone scrolling through and reading a report -- only if they are specifically inside a plot focus are those tools usable, and I'd personally much rather they not be there to distract me. In any case, I don't need to make that call for Bokeh.

> https://vega.github.io/vega-themes/

Yes, that's a nice demo and a good set of themes. For what it's worth, here are my personal notes on those themes, with a * indicating a theme I'd think is worth copying:

*default: Nice!
excel: Maybe useful to match Excel, but not my favorite
ggplot2: grim!
quartz: ok; maybe good for colorblind?
vox: grim!
*dark: Nice! Seems to have an error, though, as the colormap for the Unemployment rate increases towards the background color and should increase away from it.
fivethirtyeight: Mostly nice, but I don't love the background lines
*latimes: Nice!
urbaninstitute: Reasonably nice
googlecharts: Reasonably nice
powerbi: Ok, but confusing mix of serif and nonserif fonts
carbonwhite: Ok
carbong10: Not sure it's reasonable for a theme to change the background of the page to light grey, since there are surely non-chart things on the page too.
carbong90: Busy; also has wrong colormaps (rate increasing towards the background color)
carbong100: Busy; also has wrong colormaps (rate increasing towards the background color)
